### PR TITLE
[ONNX][PTQ] Align statistics instances with OV, Torch

### PR DESCRIPTION
**Pull request by kshpv**
### Changes

Implement statistics collection for ONNX via reducers and aggregators approach.

### Reason for changes

Align with OV and Torch.

### Related tickets

111117
121650

### Tests

All tests are updated accordingly.

### CODE REVIEW

**Review by daniil-lyakhov — CHANGES_REQUESTED** [review]
Please don't forget to implement template test for reducers and aggregators:
`tests/common/experimental/test_reducers_and_aggregators.py`

File: nncf/quantization/algorithms/min_max/onnx_backend.py [hunk 1]
```diff
@@ -151,41 +187,52 @@ def get_statistic_collector(
...
-    ) -> Union[ONNXMinMaxStatisticCollector, ONNXMeanMinMaxStatisticCollector]:
+    ) -> TensorCollector:
         is_per_channel = quantizer_config.per_channel
         node = nncf_graph.get_node_by_name(target_point.target_node_name)
         use_abs_max = quantizer_config.mode == QuantizationMode.SYMMETRIC
-        reduction_shape = None  # Per-Tensor
         quantization_axis = get_quantization_axis(is_per_channel, node, target_point)
         quantized_tensor_shape = get_quantized_tensor_shape(nncf_graph, node, target_point)
-        if quantization_axis is not None and quantized_tensor_shape is not None:  # Per-Channel
-            reduction_shape = get_reduction_shape(quantized_tensor_shape, quantization_axis)
```
[the remark this answers is not in the document]
Comment (kshpv) [reply nncf/quantization/algorithms/min_max/onnx_backend.py:162]: reverted

File: tests/onnx/quantization/test_quantizer_config.py [hunk 2]
```diff
@@ -33,14 +35,20 @@ class TestQuantizerConfig(TemplateTestQuantizerConfig):
...
+        assert MinAggregator in aggrs
+        assert MaxAggregator in aggrs
 
-    def check_is_mean_min_max_statistic_collector(self, tensor_collector):
-        assert isinstance(tensor_collector, ONNXMeanMinMaxStatisticCollector)
+    def check_is_mean_min_max_statistic_collector(self, tensor_collector: TensorCollector):
+        aggrs = [aggr.__class__ for aggr in tensor_collector.aggregators.values()]
+        assert len(aggrs) == 2
+        assert MeanAggregator in aggrs
+        assert aggrs[0].__class__ == aggrs[1].__class__
```
[the remark this answers is not in the document]
Comment (kshpv) [reply tests/onnx/quantization/test_quantizer_config.py]: done

File: nncf/common/tensor.py [hunk 3]
```diff
@@ -44,4 +44,6 @@ def device(self) -> DeviceType:
         pass
 
     def is_empty(self) -> bool:
+        if 0 in self.shape:
+            return True
         return False
```
Comment (daniil-lyakhov) [inline nncf/common/tensor.py:49]: Minor
```suggestion
       return 0 in self.shape
```

Comment (kshpv) [reply nncf/common/tensor.py]: done

File: nncf/common/tensor.py [hunk 4]
```diff
@@ -44,4 +44,4 @@ def device(self) -> DeviceType:
         pass
 
     def is_empty(self) -> bool:
-        return False
+        return 0 in self.shape
```
Comment (daniil-lyakhov) [inline nncf/common/tensor.py:47]: Maybe it is better to make this method abstract and implement this backend-specific function in each separate backend?

**Review by daniil-lyakhov — CHANGES_REQUESTED** [review]

File: nncf/common/tensor.py (hunk 4 above)
Comment (kshpv) [reply nncf/common/tensor.py]: done

Comment (kshpv) [reply nncf/common/tensor.py]: make abstract method and added implementations for ONNX and TF

File: nncf/tensorflow/tensor.py [hunk 5]
```diff
@@ -30,3 +30,6 @@ def __init__(self, tensor: tf.Tensor):
     @property
     def device(self) -> tf.device:
         return self._tensor.device
+
+    def is_empty(self) -> bool:
+        tf.equal(tf.size(self._tensor), 0)
```
Comment (daniil-lyakhov) [inline nncf/tensorflow/tensor.py:35]: Return is missed here

**Review by daniil-lyakhov — APPROVED** [review]

File: nncf/experimental/common/tensor_statistics/collectors.py [hunk 6]
```diff
@@ -31,6 +31,13 @@
 AggregationAxes = Tuple[int, ...]
 
 
+class OutputMetadata:
```
Comment (alexsu52) [inline nncf/experimental/common/tensor_statistics/collectors.py:34]: Could you briefly explain why you need to introduce the  OutputMetadata abstraction?

Comment (kshpv) [reply nncf/experimental/common/tensor_statistics/collectors.py:34]: @daniil-lyakhov

File: tests/onnx/test_statistics_aggregator.py [hunk 7]
```diff
@@ -84,18 +85,13 @@ def reducers_map(self) -> List[TensorReducerBase]:
     @pytest.fixture
     def dataset_samples(self, dataset_values):
         input_shape = INPUT_SHAPE
-        dataset_samples = [np.zeros(input_shape), np.ones(input_shape)]
+        dataset_samples = [np.zeros(input_shape).astype(np.float32), np.ones(input_shape).astype(np.float32)]
```
Comment (AlexanderDokuchaev) [inline tests/onnx/test_statistics_aggregator.py:88]:
```suggestion
        dataset_samples = [np.zeros(input_shape, dtype=np.float32), np.ones(input_shape, dtype=np.float32)]
```

File: tests/common/experimental/test_statistic_collector.py [hunk 8]
```diff
@@ -322,7 +333,7 @@ class BadStatContainer:
 
     tensor_collector = TensorCollector(BadStatContainer)
     tensor_collector.register_statistic_branch("A", DummyTensorReducer("A"), DummyTensorAggregator())
-    tensor_collector.register_input_for_all_reducers(NumpyNNCFTensor(1))
+    tensor_collector.register_input_for_all_reducers(NumpyNNCFTensor(np.array((1))))
```
Comment (AlexanderDokuchaev) [inline tests/common/experimental/test_statistic_collector.py:336]:
```suggestion
    tensor_collector.register_input_for_all_reducers(NumpyNNCFTensor(np.array(1)))
```

**Review by AlexanderDokuchaev — CHANGES_REQUESTED** [review]

File: nncf/experimental/common/tensor_statistics/collectors.py (hunk 6 above)
Comment (daniil-lyakhov) [reply nncf/experimental/common/tensor_statistics/collectors.py:34]: OV backend is using port_id and node name to specify unique output. ONNX, in other hand, is using only edge name to specify this. As this output identificatory is backend-specific, I've introduced backend-specific output metadata class

File: nncf/common/tensor.py [hunk 9]
```diff
@@ -43,5 +43,6 @@ def shape(self) -> List[int]:
     def device(self) -> DeviceType:
         pass
 
+    @abstractmethod
```
[the remark this answers is not in the document]
Comment (kshpv) [reply nncf/common/tensor.py:46]: then I will change to NotImplementedError

File: nncf/experimental/common/tensor_statistics/collectors.py [hunk 10]
```diff
@@ -74,13 +81,12 @@ def _reduce_out_of_place(self, x: List[TensorType]) -> List[TensorType]:
         """
 
     @abstractmethod
-    def get_output_names(self, target_node_name: str, port_id: int) -> List[str]:
+    def get_output_names(self, output_metadata: OutputMetadata) -> List[str]:
```
[the remark this answers is not in the document]
Comment (kshpv) [reply nncf/experimental/common/tensor_statistics/collectors.py:84]: @daniil-lyakhov

File: nncf/onnx/statistics/collectors.py [hunk 11]
```diff
@@ -157,65 +170,122 @@ def batch_mean(x: NNCFTensor) -> NNCFTensor:
...
+
+
+class ONNXBasicReducer(TensorReducerBase):
+    def _get_processor(self):
         return ONNXNNCFCollectorTensorProcessor
 
-    def _register_input(self, x: ONNXNNCFTensor):
-        self._register_input_common(x)
+    def get_inplace_fn(self):
+        raise RuntimeError("ONNX backend has no support of inplace statistics yet.")
```
[the remark this answers is not in the document]
Comment (kshpv) [reply nncf/onnx/statistics/collectors.py]: done

File: tests/onnx/test_statistics_aggregator.py (hunk 7 above)
Comment (kshpv) [reply tests/onnx/test_statistics_aggregator.py]: done

File: tests/common/experimental/test_statistic_collector.py (hunk 8 above)
Comment (kshpv) [reply tests/common/experimental/test_statistic_collector.py]: done

File: nncf/experimental/common/tensor_statistics/collectors.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (daniil-lyakhov) [reply nncf/experimental/common/tensor_statistics/collectors.py:84]: @AlexanderDokuchaev, `TensorLocator` and `TensorInfo` is too general names as this structure is primary used to locate outputs, IMHO

[the remark this answers is not in the document]
Comment (kshpv) [reply nncf/experimental/common/tensor_statistics/collectors.py]: reverted

File: nncf/experimental/common/tensor_statistics/collectors.py (hunk 6 above)
Comment (kshpv) [reply nncf/experimental/common/tensor_statistics/collectors.py]: reverted

**Review by alexsu52 — COMMENTED** [review]
Could you provide e2e and ptq tests results?

**Comment by kshpv** [comment]
> Could you provide e2e and ptq tests results?
e2e - job 543

**Review by AlexanderDokuchaev — APPROVED** [review]

**Comment by kshpv** [comment]
> Could you provide e2e and ptq tests results?

ptq conformance manual job is 252

**Comment by kshpv** [comment]
@alexsu52 during validation no regression was detected

**Review by alexsu52 — APPROVED** [review]

**Pull request merged**